proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+ */
+@property (nonatomic) SDLViewController *presentingModalViewController;
+
+/**
+ * This modal view controller is currently presented on the screen (by the presenting view controller).
+ */
+@property (nonatomic) SDLModalViewController *presentedModalViewController;
+
+/**
+ * Presents a modal view controller on the screen.
+ * Calls the completion handler if the modal view controller disappears from the screen or if an error occurred.
+ */
+- (void)presentModalViewController:(SDLModalViewController *)modalViewController completion:(void (^)(SDLResult result, NSError *error))completion;
+
+@end
+```
+
+When a view controller (presenting VC) want's to present a modal view controller (presented VC) it must call the `presentModalViewController:completion:` method of the view controller manager. The presenting VC must be listed in the view controller stack but does not need to be the top view controller. The presenting VC, presented VC and the completion handler are managed in a VC pair.
+
+The VC manager should initiate the presentation (send the RPC request) and store the VC pair in a private list. It should be possible to associate RPC responses to the VC pair in the list. If a response is received and can be associated to a VC pair then the VC pair should be removed from the list and the completion handler should be called.
+
+The oldest VC pair of this list should be the currently presenting and presented VC. The currently presenting VC will be accessible through the `.presentingModalViewController` property and the presented VC through the `.presentedModalViewController`. If the private list is empty the two properties should return nil.
+
+By the nature of the HMI overlays can be aborted by another overlay or get rejected because of a currently presented overlay (or something of a higher priority e.g. VR session). The private list offers flexibility in case the app calls the method multiple times and can deal with rejected overlays that were never visible and aborted overlays if another one is requested.
+
+#### SDLModalViewController
+
+This class should be used as the base class for any kind of overlay. As it inherits `SDLViewController` it is possible to manipulate the head unit during a modal presentation.
+
+```objc
+@interface SDLModalViewController : SDLViewController
+
+@property (nonatomic, assign) NSTimeInterval duration; // used throughout all subclasses
+@property (nonatomic, copy) NSString *message; // used throughout all subclasses
+
+- (void)present:(SDLApplication *)application; // private method called by the VC manager
+
+@end
+```
+
+#### SDLAlertController
+
+The alert controller is used to abstract the RPC `Alert`. The properties match the parameters of the RPC.
+
+```objc
+@interface SDLAlertController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
```

Comment:
This should probably be `SDLSoftButtonObject` so that developers don't have to deal with `buttonId` and can auto-upload an image?